src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs

```diff
@@ -271,6 +285,9 @@ private async Task InjectNewHttp3ConnectionAsync(RequestQueue<Http3Connection?>.
                     CreateConnectTimeoutException(oce) :
                     e;
 
+                Debug.Assert(connectionSetupActivity?.IsStopped is not true);
```

Comment:
In that case we would stop the Activity after calling `connection.InitQuicConnection()` / `MarkConnectionAsEstablished`. This might be harmless, but would be inconsistent with H/1.1 H/2 behavior, which I want to avoid.

Reply:
Note that we call `StopConnectionSetupActivity()` at L274. It does more than just stopping the activity (not prepared for double-stopping) and I prefer to do the stop on the success path before the `InitQuicConnection() -> MarkConnetionAsEstablished()` calls in order to be consistent with H/1.1 H/2. I wanted to avoid defining a special method in `ConnectionSetupDiagnostics` that would check for `Activity.IsStopped` just for the H3 case.

I guess this logic deserves more comments?

Reply:
I see, yeah, if you could mention that the stopping should be done before the other diagnostics work is done in `MarkConnetionAsEstablished` (via `InitQuicConnection`) to be consistent with H/1.1 H/2.